spopt/locate/p_median.py

```diff
@@ -533,3 +535,576 @@ def solve(self, solver: pulp.LpSolver, results: bool = True):
+
+        Raises
+        ------
+        ValueError
+            If any value in the k_list is greater than the total number of facilities.
+
+        Notes
+        -----
+        This method uses a suitable tree data structure (built with the
+        `build_best_tree` function). To efficiently find the k nearest
+        facilities for each client based on the specified distance metric.
+        The resulting distances are stored in a sparse matrix format to
+        conserve memory for large datasets.
+        """
+
+        row_shape = len(clients)
+        column_shape = len(facilities)
+
+        # check the k value with the total number of facilities
+        for k in k_list:
+            if k > column_shape:
+                raise ValueError(
+                    f"The value of k should be no more than the number of total"
+                    f"facilities ({column_shape})."
+                )
+
+        # Initialize empty lists to store the data for the sparse matrix
+        data = []
+        row_index = []
+        col_index = []
+
+        # create the suitable Tree
+        tree = build_best_tree(facilities, metric)
+
+        for i, k in enumerate(k_list):
+            # Query the Tree to find the k nearest facilities for each client
+            distances, k_nearest_facilities_indices = tree.query([clients[i]], k=k)
+
+            # extract the contents of the inner array
+            distances = distances[0].tolist()
+            k_nearest_facilities_indices = k_nearest_facilities_indices[0].tolist()
+
+            # Append the data for the sparse matrix
+            data.extend(distances)
+            row_index.extend([i] * k)
+            col_index.extend(k_nearest_facilities_indices)
+        # Create the sparse matrix using csr_matrix
+        sparse_matrix = csr_matrix(
+            (data, (row_index, col_index)), shape=(row_shape, column_shape)
+        )
+        return sparse_matrix
+
+    def _create_k_list(self, k_list: np.array):
+        """
+        Increase the k value of clients with any g_i > 0 and create a new k list.
+
+        Parameters
+        ----------
+        k_list : np.array
+            An array of integers representing the original k values for each client.
+
+        Returns
+        -------
+        new_k_list (np.array)
+            A new array of integers with increased k values for clients with g_i > 0.
+
+        Notes
+        -----
+        This method is used to adjust the k values for clients based on
+        their placeholder variable g_i. For clients with g_i greater than 0,
+        the corresponding k value is increased by 1 in the new k list.
+        """
+
+        new_k_list = k_list.copy()
+        placeholder_vars = getattr(self, "placeholder_vars")
+        for i in range(len(placeholder_vars)):
+            if placeholder_vars[i].value() > 0:
+                new_k_list[i] = new_k_list[i] + 1
+        return new_k_list
+
+    @classmethod
```

Comment:
this `@classmethod` needs to be removed. And, at the end of the function, we would set `self.problem = k_nearest_p_median`